File: cpp/repl/big-numbers2/parser.cpp
#include "parser.h"

Parser::Parser() {

}

Parser::~Parser() {

}

void Parser::load(std::string source) {
    lexer.load(source);
    nextToken();
    nextToken();
}

void Parser::nextToken() {
    currentToken = peekToken;
    peekToken = lexer.nextToken();
}

void Parser::buildProgramStatements(AstProgram program) {
    while(currentToken.getType() != TT_EOF) {
        std::cout << currentToken.toString() << "\n";
        AstNode* stmt = parseStatement();
        if (stmt) program.statements.push_back(stmt);
        nextToken();
    }
}

AstNode* Parser::parseStatement() {
    if (currentToken.getType == TT_LET) {
        //... return parseLetStatement();
    }
    //... return ParseExpressionStatement();
    return NULL;
}

File: java/VMPrototype1/src/main/java/ro/adrianus/prototype/parser/Parser.java
package ro.adrianus.prototype.parser;

import ro.adrianus.prototype.ast.AstNode;
import ro.adrianus.prototype.ast.AstProgram;
import ro.adrianus.prototype.lexer.Lexer;
import ro.adrianus.prototype.token.Token;
import ro.adrianus.prototype.token.TokenTypeEnum;

import java.util.HashMap;
import java.util.function.Function;
import java.util.function.Supplier;

public class Parser {
    private Lexer lexer;
    private String source;
    private Token currentToken;
    private Token peekToken;
    private HashMap<TokenTypeEnum, Supplier<AstNode>> prefixParseFns = new HashMap<>();
    private HashMap<TokenTypeEnum, Function<AstNode, AstNode>> infixParseFns = new HashMap<>();

    public Parser() {
        lexer = new Lexer();
        prefixParseFns.put(TokenTypeEnum.SUB, this::parsePrefixExpression);
        infixParseFns.put(TokenTypeEnum.PLUS, this::parseInfixExpression);
        infixParseFns.put(TokenTypeEnum.SUB, this::parseInfixExpression);
        infixParseFns.put(TokenTypeEnum.MULTIPLY, this::parseInfixExpression);
        infixParseFns.put(TokenTypeEnum.DIVIDE, this::parseInfixExpression);
    }

    public void load(String s) {
        source = s;
        lexer.load(source);
        nextToken();
        nextToken();
    }

    public void buildProgramStatements(AstProgram program) {
        while(currentToken.getType() != TokenTypeEnum.EOF) {

        }
    }

    public void nextToken() {
        currentToken = peekToken;
        currentToken = lexer.nextToken();
    }

    AstNode parsePrefixExpression() {
        return null;
    }

    AstNode parseInfixExpression(AstNode left) {
        return null;
    }
}
